File: Python3/프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
from collections import deque
from collections import Counter

def solution(m, n, board):
    answer = 0
    board, dq = [list(b) for b in board], deque()
    r, c, flag = 0, 0, False
    
    def print_board():
        for i in range(m):
            for j in range(n):
                print(board[i][j], end="\t")
            print()
        print()
    
    while True:
        flag = False
        # 1. 겹치는 블록 찾기
        for i in range(m - 1):
            for j in range(n - 1):
                if [board[i + 1][j], board[i][j + 1], board[i + 1][j + 1]].count(board[i][j]) == 3:
                    dq.extend([[i, j], [i + 1, j], [i, j + 1], [i + 1, j + 1]])
                    flag = True
        
        if not flag:
            break
        # 2. 겹치는 블록 제거
        while dq:
            r, c = dq.pop()
            if board[r][c].isalpha():
                board[r][c] = str(r) + "000" + str(c)
                answer += 1
                # print_board()
        #3. 블록 내리기
        while flag:
            flag = False
            for i in range(m - 1):
                for j in range(n):
                    if board[i][j].isalpha() and board[i + 1][j].isdigit():
                        # print(board[i][j], "isalpha and", board[i + 1][j], "isdigit")
                        board[i][j], board[i + 1][j] = board[i + 1][j], board[i][j]
                        flag = True
        # print_board()
            
    
    return answer

File: C++/프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int solution(int m, int n, vector<string> board) {
    int answer = 0;
    queue<pair<int, int>> q;
    bool visited[m][n];
    pair<int, int> el;
    vector<int> dr = {0, 1, 0, 1}, dc = {0, 0, 1, 1};
    bool flag;
    
    while(true){
        flag = false;
        // 1. 겹치는 블록 찾기
        for(int i = 0; i < m - 1; i++) {
            for(int j = 0; j < n - 1; j++) {
                if(board[i][j] != '1' &&board[i][j] == board[i + 1][j] && board[i][j] == board[i][j + 1]
                   && board[i][j] == board[i + 1][j + 1]) {
                    flag = true;
                    for(int k = 0; k < 4; k++) {
                        q.push(make_pair(i + dr[k], j + dc[k]));
                    }
                }
            }
        }
        
        if (!flag) break;

        // 2. 겹치는 블록 제거
        while(!q.empty()) {
            el = q.front();
            q.pop();
            if(board[el.first][el.second] != '1') {
                board[el.first][el.second] = '1';
                answer++;
            }
        }

        // 3. 블록 밑으로 내리기
        char tmp;
        for(int rep = 0; rep < m; rep++) {
            for(int i = 0; i < m - 1; i++) {
                for(int j = 0; j < n; j++) {
                    if(board[i][j] != '1' && board[i + 1][j] == '1') {
                        tmp = board[i][j];
                        board[i][j] = board[i + 1][j];
                        board[i + 1][j] = tmp;
                    }
                }
            }
        }
    }
    
    
    return answer;
}